{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Transcriptome Reconstruction\n",
    "This tutorial processes the example data set, based on PacBio Isoseq samples of hematopoetic cells from ENCODE. This dataset contains only a subset of genomic regions, allowing for fast processing of the demonstration tutorials. \n",
    "All required data files to run the tutorials can be obtained here: ([download link](https://nc.molgen.mpg.de/cloud/index.php/s/zYe7g6qnyxGDxRd)). \n",
    "\n",
    "To run this example, you will you will need the following **input files**:\n",
    "\n",
    "* sample description file 'encode_samples.tsv'\n",
    "* six .bam alignment files for the six samples\n",
    "* six corresponding .bam.bai indices\n",
    "* reference annotation file gencode.v42.chr_patch_hapl_scaff.annotation_sorted_chr8.gff3.gz\n",
    "* corresponding .gff3.gz.tbi index file\n",
    "* genomic reference file GRCh38.p13.genome_chr8.fa\n",
    "* corresponding .fai index file\n",
    "\n",
    "All files are assumed to be stored in a subfolder called 'demonstration_dataset'\n",
    "\n",
    "In this tutorial, we import the reference annotation, specify the alignment files for the samples, and integrate the data into a common data structure. During this step, the transcriptome is reconstructed, and quality control metrics are computed. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:This is isotools version 0.3.5rc10\n"
     ]
    }
   ],
   "source": [
    "# preparation: import the libraries\n",
    "from isotools import Transcriptome\n",
    "from isotools import __version__ as isotools_version\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import logging\n",
    "# set up logging\n",
    "logging.basicConfig(format='%(levelname)s:%(message)s', level=logging.INFO)\n",
    "logger=logging.getLogger('isotools')\n",
    "logger.info(f'This is isotools version {isotools_version}')\n",
    "\n",
    "\n",
    "path='demonstration_dataset'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import of reference annotation\n",
    "The first step is to import the reference annotation from a gff or gtf file with the [Transcriptome.from_reference](../isotoolsAPI.html?highlight=from_reference#isotools.Transcriptome.from_reference) class method. The input file should be sorted and indexed with tabix. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:importing reference from gff3 file demonstration_dataset/gencode.v42.chr_patch_hapl_scaff.annotation_sorted_chr8.gff3.gz\n",
      "100%|█████████▉| 2.70M/2.70M [00:02<00:00, 1.22MB/s]\n",
      "INFO:skipped the following categories: {'three_prime_UTR', 'five_prime_UTR', 'CDS'}\n"
     ]
    }
   ],
   "source": [
    "annotation_fn=f'{path}/gencode.v42.chr_patch_hapl_scaff.annotation_sorted_chr8.gff3.gz'\n",
    "#create the IsoTools transcriptome object from the reference annotation\n",
    "isoseq=Transcriptome.from_reference(annotation_fn)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import of sequencing data\n",
    "Next a sample table containing the sample names as well as corresponding file names and group assignment is imported.\n",
    "This information is used in the next step, to reconstruct the transcripts for each sample and integrate them with the transcriptome object, using the [add_sample_from_bam](../isotoolsAPI.html?highlight=add_sample_from_bam#isotools.Transcriptome.add_sample_from_bam) function.\n",
    "When all samples are added, quality control metrics are calculated by calling [Transcriptome.add_qc_metrics](../isotoolsAPI.html?highlight=add_qc_metrics#isotools.Transcriptome.add_qc_metrics).\n",
    "Last, the object with all data is stored on disk in a pickle file for later use with the [save](../isotoolsAPI.html?highlight=save#isotools.Transcriptome.save) method.\n",
    "In another session, the object can be restored with the classmethod [Transcriptome.load](../isotoolsAPI.html?highlight=load#isotools.Transcriptome.load)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>file_name</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GM12878_a</td>\n",
       "      <td>demonstration_dataset/ENCFF417VHJ_aligned_mm2_...</td>\n",
       "      <td>GM12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GM12878_b</td>\n",
       "      <td>demonstration_dataset/ENCFF450VAU_aligned_mm2_...</td>\n",
       "      <td>GM12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GM12878_c</td>\n",
       "      <td>demonstration_dataset/ENCFF694DIE_aligned_mm2_...</td>\n",
       "      <td>GM12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>K562_a</td>\n",
       "      <td>demonstration_dataset/ENCFF696GDL_aligned_mm2_...</td>\n",
       "      <td>K562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>K562_b</td>\n",
       "      <td>demonstration_dataset/ENCFF429VVB_aligned_mm2_...</td>\n",
       "      <td>K562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>K562_c</td>\n",
       "      <td>demonstration_dataset/ENCFF634YSN_aligned_mm2_...</td>\n",
       "      <td>K562</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name                                          file_name    group\n",
       "0   GM12878_a  demonstration_dataset/ENCFF417VHJ_aligned_mm2_...  GM12878\n",
       "1   GM12878_b  demonstration_dataset/ENCFF450VAU_aligned_mm2_...  GM12878\n",
       "2   GM12878_c  demonstration_dataset/ENCFF694DIE_aligned_mm2_...  GM12878\n",
       "3      K562_a  demonstration_dataset/ENCFF696GDL_aligned_mm2_...     K562\n",
       "4      K562_b  demonstration_dataset/ENCFF429VVB_aligned_mm2_...     K562\n",
       "5      K562_c  demonstration_dataset/ENCFF634YSN_aligned_mm2_...     K562"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_fn=f'{path}/encode_samples.tsv'\n",
    "genome_fn=f'{path}/GRCh38.p13.genome_chr8.fa'\n",
    "\n",
    "samples=pd.read_csv(sample_fn, sep='\\t')\n",
    "samples.file_name=path+'/'+samples.file_name\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:adding sample GM12878_a from file demonstration_dataset/ENCFF417VHJ_aligned_mm2_chr8.bam\n",
      "100%|██████████| 53.0k/53.0k [00:12<00:00, 4.12kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 113 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 10940 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "WARNING:ignored 532 chimeric alignments with only one part aligned to specified chromosomes.\n",
      "INFO:ignoring 2235 chimeric alignments with less than 2 reads\n",
      "INFO:imported 40177 nonchimeric reads (including  14 chained chimeric alignments) and 73 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample GM12878_b from file demonstration_dataset/ENCFF450VAU_aligned_mm2_chr8.bam\n",
      "100%|██████████| 68.3k/68.3k [00:13<00:00, 5.16kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 72 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 12598 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "WARNING:ignored 485 chimeric alignments with only one part aligned to specified chromosomes.\n",
      "INFO:ignoring 1275 chimeric alignments with less than 2 reads\n",
      "INFO:imported 54856 nonchimeric reads (including  12 chained chimeric alignments) and 7 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample GM12878_c from file demonstration_dataset/ENCFF694DIE_aligned_mm2_chr8.bam\n",
      "[W::hts_idx_load3] The index file is older than the data file: demonstration_dataset/ENCFF694DIE_aligned_mm2_chr8.bam.bai\n",
      "100%|██████████| 90.6k/90.6k [00:15<00:00, 6.03kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 84 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 17108 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "WARNING:ignored 456 chimeric alignments with only one part aligned to specified chromosomes.\n",
      "INFO:ignoring 1412 chimeric alignments with less than 2 reads\n",
      "INFO:imported 72445 nonchimeric reads (including  38 chained chimeric alignments) and 12 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample K562_a from file demonstration_dataset/ENCFF696GDL_aligned_mm2_chr8.bam\n",
      "[W::hts_idx_load3] The index file is older than the data file: demonstration_dataset/ENCFF696GDL_aligned_mm2_chr8.bam.bai\n",
      "100%|██████████| 77.9k/77.9k [00:16<00:00, 4.61kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 166 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 14966 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "WARNING:ignored 1145 chimeric alignments with only one part aligned to specified chromosomes.\n",
      "INFO:ignoring 4547 chimeric alignments with less than 2 reads\n",
      "INFO:imported 59121 nonchimeric reads (including  49 chained chimeric alignments) and 281 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample K562_b from file demonstration_dataset/ENCFF429VVB_aligned_mm2_chr8.bam\n",
      "100%|██████████| 107k/107k [00:20<00:00, 5.25kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 299 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 23934 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "WARNING:ignored 2161 chimeric alignments with only one part aligned to specified chromosomes.\n",
      "INFO:ignoring 7465 chimeric alignments with less than 2 reads\n",
      "INFO:imported 76686 nonchimeric reads (including  63 chained chimeric alignments) and 415 chimeric reads with coverage of at least 2.\n",
      "INFO:adding sample K562_c from file demonstration_dataset/ENCFF634YSN_aligned_mm2_chr8.bam\n",
      "[W::hts_idx_load3] The index file is older than the data file: demonstration_dataset/ENCFF634YSN_aligned_mm2_chr8.bam.bai\n",
      "100%|██████████| 117k/117k [00:22<00:00, 5.10kreads/s, chr=KI270757.1]\n",
      "INFO:skipped 295 reads aligned fraction of less than 0.75.\n",
      "INFO:skipped 30149 secondary alignments (0x100), alignment that failed quality check (0x200) or PCR duplicates (0x400)\n",
      "WARNING:ignored 2522 chimeric alignments with only one part aligned to specified chromosomes.\n",
      "INFO:ignoring 8023 chimeric alignments with less than 2 reads\n",
      "INFO:imported 80338 nonchimeric reads (including  46 chained chimeric alignments) and 369 chimeric reads with coverage of at least 2.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>file</th>\n",
       "      <th>group</th>\n",
       "      <th>nonchimeric_reads</th>\n",
       "      <th>chimeric_reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GM12878_a</td>\n",
       "      <td>demonstration_dataset/ENCFF417VHJ_aligned_mm2_...</td>\n",
       "      <td>GM12878</td>\n",
       "      <td>40177</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GM12878_b</td>\n",
       "      <td>demonstration_dataset/ENCFF450VAU_aligned_mm2_...</td>\n",
       "      <td>GM12878</td>\n",
       "      <td>54856</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GM12878_c</td>\n",
       "      <td>demonstration_dataset/ENCFF694DIE_aligned_mm2_...</td>\n",
       "      <td>GM12878</td>\n",
       "      <td>72445</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>K562_a</td>\n",
       "      <td>demonstration_dataset/ENCFF696GDL_aligned_mm2_...</td>\n",
       "      <td>K562</td>\n",
       "      <td>59121</td>\n",
       "      <td>281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>K562_b</td>\n",
       "      <td>demonstration_dataset/ENCFF429VVB_aligned_mm2_...</td>\n",
       "      <td>K562</td>\n",
       "      <td>76686</td>\n",
       "      <td>415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>K562_c</td>\n",
       "      <td>demonstration_dataset/ENCFF634YSN_aligned_mm2_...</td>\n",
       "      <td>K562</td>\n",
       "      <td>80338</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        name                                               file    group  \\\n",
       "0  GM12878_a  demonstration_dataset/ENCFF417VHJ_aligned_mm2_...  GM12878   \n",
       "1  GM12878_b  demonstration_dataset/ENCFF450VAU_aligned_mm2_...  GM12878   \n",
       "2  GM12878_c  demonstration_dataset/ENCFF694DIE_aligned_mm2_...  GM12878   \n",
       "3     K562_a  demonstration_dataset/ENCFF696GDL_aligned_mm2_...     K562   \n",
       "4     K562_b  demonstration_dataset/ENCFF429VVB_aligned_mm2_...     K562   \n",
       "5     K562_c  demonstration_dataset/ENCFF634YSN_aligned_mm2_...     K562   \n",
       "\n",
       "  nonchimeric_reads chimeric_reads  \n",
       "0             40177             73  \n",
       "1             54856              7  \n",
       "2             72445             12  \n",
       "3             59121            281  \n",
       "4             76686            415  \n",
       "5             80338            369  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# integrate the samples\n",
    "for i,row in samples.iterrows():\n",
    "    # this step takes about 5-30 seconds per sample\n",
    "     isoseq.add_sample_from_bam(row.file_name, sample_name=row.sample_name, group=row.group)\n",
    "# the sample table of the transcriptome object contains the number of imported reads\n",
    "isoseq.sample_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the next step, we compute several qc metrics for the transcripts: \n",
    "\n",
    "* downstream A content, \n",
    "* direct repeat length at junctions, \n",
    "* noncanonical splicing, \n",
    "* potential fragments\n",
    "\n",
    "This information is stored with the genes objects, and can be accessed by downstream analysis. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 10801/10801 [01:06<00:00, 162.33genes/s]\n",
      "100%|██████████| 10801/10801 [00:49<00:00, 220.16genes/s]\n"
     ]
    }
   ],
   "source": [
    "# compute qc metrics\n",
    "isoseq.add_qc_metrics(genome_fn)\n",
    "# add ORF predictions\n",
    "isoseq.add_orf_prediction(genome_fn)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Transcriptome Export\n",
    "The identified transcrips can be exported in pickle format, GTF format, and as a table.\n",
    "* Pickle format is an internal python format, used to save the entire transcriptome data, so it can be restored in an IsoTools session without reimporting the alignment files. Export to pickle is done with the [save](../isotoolsAPI.html?highlight=save#isotools.Transcriptome.save) methods\n",
    "* GTF (General Feature Format) is a file format commonly used in bioinformatics to represent annotated genomic features, such as gene models, exons, and introns. Exporting to this format is done with the [write_gtf](../isotoolsAPI.html?highlight=write_gtf#isotools.Transcriptome.write_gtf) method, and facilitates the use of the reconstructed transcriptome in several external tools. \n",
    "* The transcripts can also be exported as a table, containing coverage information, as well as additional features such as qc metrics, specified with the \"extra_columns\" parameter. The use of the [transcript_table](../isotoolsAPI.html?highlight=transcript_table#isotools.Transcriptome.transcript_table) methods is explained in the API documentation. \n",
    "\n",
    "\n",
    "To select the transcripts to be exported to GTF and the table, we used the isotools filtering query syntax, which is explained in detail in a [separate tutorial](06_filtering.html#Filtering-tags-and-queries). \n",
    "The gtf can be compressed by adding a '.gz' to the filename and setting  gzip=True. \n",
    "The table can be set to contain samplewise and/or groupwise coverage information, by setting the sample and the groups parameters. Here we sum the read counts per group. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:saving transcriptome to demonstration_dataset/PacBio_isotools.pkl\n"
     ]
    }
   ],
   "source": [
    "# export the transcriptome object for later use.\n",
    "isoseq.save(f'{path}/PacBio_isotools.pkl')\n",
    "# to load the data in the next session, use\n",
    "# isoseq=Transcriptome.load('PacBio_isotools_substantial_isotools.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:writing gzip compressed gtf file to demonstration_dataset/demonstration_dataset_substantial_transcripts.gtf.gz\n"
     ]
    }
   ],
   "source": [
    "# export gtf:\n",
    "query_string = 'SUBSTANTIAL and not (NOVEL_TRANSCRIPT and UNSPLICED)'\n",
    "isoseq.write_gtf(f'{path}/demonstration_dataset_substantial_transcripts.gtf.gz',\n",
    "                 source='isoseq', min_coverage=5,\n",
    "                 gzip=True,\n",
    "                 query=query_string)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 10801/10801 [00:00<00:00, 22686.76genes/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>transcript_start</th>\n",
       "      <th>transcript_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_nr</th>\n",
       "      <th>transcript_length</th>\n",
       "      <th>num_exons</th>\n",
       "      <th>exon_starts</th>\n",
       "      <th>exon_ends</th>\n",
       "      <th>novelty_class</th>\n",
       "      <th>novelty_subclasses</th>\n",
       "      <th>GM12878_sum_coverage</th>\n",
       "      <th>K562_sum_coverage</th>\n",
       "      <th>GM12878_sum_tpm</th>\n",
       "      <th>K562_sum_tpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr8</td>\n",
       "      <td>15540261</td>\n",
       "      <td>15764645</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000104723.21</td>\n",
       "      <td>TUSC3</td>\n",
       "      <td>16</td>\n",
       "      <td>1640</td>\n",
       "      <td>10</td>\n",
       "      <td>15540261,15623079,15650696,15659506,15662155,1...</td>\n",
       "      <td>15540568,15623249,15650814,15659647,15662296,1...</td>\n",
       "      <td>FSM</td>\n",
       "      <td>FSM</td>\n",
       "      <td>0</td>\n",
       "      <td>226</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1045.594393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8</td>\n",
       "      <td>15540261</td>\n",
       "      <td>15764645</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000104723.21</td>\n",
       "      <td>TUSC3</td>\n",
       "      <td>17</td>\n",
       "      <td>1705</td>\n",
       "      <td>11</td>\n",
       "      <td>15540261,15623079,15650696,15659506,15662155,1...</td>\n",
       "      <td>15540568,15623249,15650814,15659647,15662296,1...</td>\n",
       "      <td>FSM</td>\n",
       "      <td>FSM</td>\n",
       "      <td>0</td>\n",
       "      <td>52</td>\n",
       "      <td>0.0</td>\n",
       "      <td>240.579241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>15540261</td>\n",
       "      <td>15764645</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000104723.21</td>\n",
       "      <td>TUSC3</td>\n",
       "      <td>18</td>\n",
       "      <td>1549</td>\n",
       "      <td>9</td>\n",
       "      <td>15540261,15623079,15650696,15659506,15662155,1...</td>\n",
       "      <td>15540568,15623249,15650814,15659647,15662296,1...</td>\n",
       "      <td>FSM</td>\n",
       "      <td>FSM</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>27.759143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr8</td>\n",
       "      <td>15540261</td>\n",
       "      <td>15758273</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000104723.21</td>\n",
       "      <td>TUSC3</td>\n",
       "      <td>20</td>\n",
       "      <td>1441</td>\n",
       "      <td>10</td>\n",
       "      <td>15540261,15623079,15650696,15659506,15662155,1...</td>\n",
       "      <td>15540568,15623249,15650814,15659647,15662296,1...</td>\n",
       "      <td>NIC</td>\n",
       "      <td>exon skipping</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0.0</td>\n",
       "      <td>92.530477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr8</td>\n",
       "      <td>15540261</td>\n",
       "      <td>15758273</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000104723.21</td>\n",
       "      <td>TUSC3</td>\n",
       "      <td>21</td>\n",
       "      <td>1506</td>\n",
       "      <td>11</td>\n",
       "      <td>15540261,15623079,15650696,15659506,15662155,1...</td>\n",
       "      <td>15540568,15623249,15650814,15659647,15662296,1...</td>\n",
       "      <td>NIC</td>\n",
       "      <td>novel combination</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0.0</td>\n",
       "      <td>92.530477</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  transcript_start  transcript_end strand             gene_id  \\\n",
       "0  chr8          15540261        15764645      +  ENSG00000104723.21   \n",
       "1  chr8          15540261        15764645      +  ENSG00000104723.21   \n",
       "2  chr8          15540261        15764645      +  ENSG00000104723.21   \n",
       "3  chr8          15540261        15758273      +  ENSG00000104723.21   \n",
       "4  chr8          15540261        15758273      +  ENSG00000104723.21   \n",
       "\n",
       "  gene_name  transcript_nr  transcript_length  num_exons  \\\n",
       "0     TUSC3             16               1640         10   \n",
       "1     TUSC3             17               1705         11   \n",
       "2     TUSC3             18               1549          9   \n",
       "3     TUSC3             20               1441         10   \n",
       "4     TUSC3             21               1506         11   \n",
       "\n",
       "                                         exon_starts  \\\n",
       "0  15540261,15623079,15650696,15659506,15662155,1...   \n",
       "1  15540261,15623079,15650696,15659506,15662155,1...   \n",
       "2  15540261,15623079,15650696,15659506,15662155,1...   \n",
       "3  15540261,15623079,15650696,15659506,15662155,1...   \n",
       "4  15540261,15623079,15650696,15659506,15662155,1...   \n",
       "\n",
       "                                           exon_ends novelty_class  \\\n",
       "0  15540568,15623249,15650814,15659647,15662296,1...           FSM   \n",
       "1  15540568,15623249,15650814,15659647,15662296,1...           FSM   \n",
       "2  15540568,15623249,15650814,15659647,15662296,1...           FSM   \n",
       "3  15540568,15623249,15650814,15659647,15662296,1...           NIC   \n",
       "4  15540568,15623249,15650814,15659647,15662296,1...           NIC   \n",
       "\n",
       "  novelty_subclasses  GM12878_sum_coverage  K562_sum_coverage  \\\n",
       "0                FSM                     0                226   \n",
       "1                FSM                     0                 52   \n",
       "2                FSM                     0                  6   \n",
       "3      exon skipping                     0                 20   \n",
       "4  novel combination                     0                 20   \n",
       "\n",
       "   GM12878_sum_tpm  K562_sum_tpm  \n",
       "0              0.0   1045.594393  \n",
       "1              0.0    240.579241  \n",
       "2              0.0     27.759143  \n",
       "3              0.0     92.530477  \n",
       "4              0.0     92.530477  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# export transcript table with the same filter criteria:\n",
    "transcript_tab=isoseq.transcript_table( groups=isoseq.groups(),tpm=True,coverage=True,\n",
    "                                       min_coverage=5, progress_bar=True,\n",
    "                                       query=query_string)\n",
    "# write to csv file\n",
    "transcript_tab.to_csv(f'{path}/demonstration_dataset_substantial_transcripts.csv',\n",
    "                      index=False, sep='\\t')\n",
    "#show the first lines\n",
    "transcript_tab.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  },
  "vscode": {
   "interpreter": {
    "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
