{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# File Export\n",
    "The identified transcrips can be exported in gtf format or as a table with additional information, with \n",
    "the [write_gtf](../isotoolsAPI.html?highlight=write_gtf#isotools.Transcriptome.write_gtf) \n",
    "and [transcript_table](../isotoolsAPI.html?highlight=transcript_table#isotools.Transcriptome.transcript_table) methods. Both methods accept all filtering functionality from iter_transcripts, allowing for flexible and fine grained filtering of the relevant transcripts. \n",
    "\n",
    "This tutorial demonstrates the export functionality with the prepared transcriptome .pkl file from [here](https://oc-molgen.gnz.mpg.de/owncloud/s/gjG9EPiQwpRAyg3)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "This is isotools version 0.3.2rc6, but data has been pickled with version 0.3.2rc2, which may be incompatible\n"
     ]
    }
   ],
   "source": [
    "from isotools import Transcriptome\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "path='demonstration_dataset'\n",
    "isoseq=Transcriptome.load(f'{path}/PacBio_isotools_substantial_isotools.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export gtf:\n",
    "isoseq.write_gtf(f'{path}/demonstration_dataset_substantial_transcripts.gtf.gz', source='isoseq', min_coverage=5,  gzip=True, query='SUBSTANTIAL and not (NOVEL_TRANSCRIPT and UNSPLICED)')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 10803/10803 [00:00<00:00, 24358.25genes/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>transcript_start</th>\n",
       "      <th>transcript_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_nr</th>\n",
       "      <th>transcript_length</th>\n",
       "      <th>num_exons</th>\n",
       "      <th>exon_starts</th>\n",
       "      <th>exon_ends</th>\n",
       "      <th>novelty_class</th>\n",
       "      <th>novelty_subclasses</th>\n",
       "      <th>GM12878_sum_coverage</th>\n",
       "      <th>K562_sum_coverage</th>\n",
       "      <th>GM12878_sum_tpm</th>\n",
       "      <th>K562_sum_tpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr8</td>\n",
       "      <td>15540262</td>\n",
       "      <td>15764483</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000104723.21</td>\n",
       "      <td>TUSC3</td>\n",
       "      <td>17</td>\n",
       "      <td>1477</td>\n",
       "      <td>10</td>\n",
       "      <td>15540262,15623079,15650696,15659506,15662155,1...</td>\n",
       "      <td>15540568,15623249,15650814,15659647,15662296,1...</td>\n",
       "      <td>FSM</td>\n",
       "      <td>FSM</td>\n",
       "      <td>0</td>\n",
       "      <td>226</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1045.555694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8</td>\n",
       "      <td>15540262</td>\n",
       "      <td>15758274</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000104723.21</td>\n",
       "      <td>TUSC3</td>\n",
       "      <td>18</td>\n",
       "      <td>1506</td>\n",
       "      <td>11</td>\n",
       "      <td>15540262,15623079,15650696,15659506,15662155,1...</td>\n",
       "      <td>15540568,15623249,15650814,15659647,15662296,1...</td>\n",
       "      <td>NIC</td>\n",
       "      <td>novel combination</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0.0</td>\n",
       "      <td>92.527053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>15540262</td>\n",
       "      <td>15764483</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000104723.21</td>\n",
       "      <td>TUSC3</td>\n",
       "      <td>19</td>\n",
       "      <td>1542</td>\n",
       "      <td>11</td>\n",
       "      <td>15540262,15623079,15650696,15659506,15662155,1...</td>\n",
       "      <td>15540568,15623249,15650814,15659647,15662296,1...</td>\n",
       "      <td>FSM</td>\n",
       "      <td>FSM</td>\n",
       "      <td>0</td>\n",
       "      <td>52</td>\n",
       "      <td>0.0</td>\n",
       "      <td>240.570337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr8</td>\n",
       "      <td>15540262</td>\n",
       "      <td>15758274</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000104723.21</td>\n",
       "      <td>TUSC3</td>\n",
       "      <td>21</td>\n",
       "      <td>1441</td>\n",
       "      <td>10</td>\n",
       "      <td>15540262,15623079,15650696,15659506,15662155,1...</td>\n",
       "      <td>15540568,15623249,15650814,15659647,15662296,1...</td>\n",
       "      <td>NIC</td>\n",
       "      <td>exon skipping</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0.0</td>\n",
       "      <td>92.527053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr8</td>\n",
       "      <td>15540262</td>\n",
       "      <td>15764646</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000104723.21</td>\n",
       "      <td>TUSC3</td>\n",
       "      <td>26</td>\n",
       "      <td>1549</td>\n",
       "      <td>9</td>\n",
       "      <td>15540262,15623079,15650696,15659506,15662155,1...</td>\n",
       "      <td>15540568,15623249,15650814,15659647,15662296,1...</td>\n",
       "      <td>FSM</td>\n",
       "      <td>FSM</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>27.758116</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  transcript_start  transcript_end strand             gene_id  \\\n",
       "0  chr8          15540262        15764483      +  ENSG00000104723.21   \n",
       "1  chr8          15540262        15758274      +  ENSG00000104723.21   \n",
       "2  chr8          15540262        15764483      +  ENSG00000104723.21   \n",
       "3  chr8          15540262        15758274      +  ENSG00000104723.21   \n",
       "4  chr8          15540262        15764646      +  ENSG00000104723.21   \n",
       "\n",
       "  gene_name  transcript_nr  transcript_length  num_exons  \\\n",
       "0     TUSC3             17               1477         10   \n",
       "1     TUSC3             18               1506         11   \n",
       "2     TUSC3             19               1542         11   \n",
       "3     TUSC3             21               1441         10   \n",
       "4     TUSC3             26               1549          9   \n",
       "\n",
       "                                         exon_starts  \\\n",
       "0  15540262,15623079,15650696,15659506,15662155,1...   \n",
       "1  15540262,15623079,15650696,15659506,15662155,1...   \n",
       "2  15540262,15623079,15650696,15659506,15662155,1...   \n",
       "3  15540262,15623079,15650696,15659506,15662155,1...   \n",
       "4  15540262,15623079,15650696,15659506,15662155,1...   \n",
       "\n",
       "                                           exon_ends novelty_class  \\\n",
       "0  15540568,15623249,15650814,15659647,15662296,1...           FSM   \n",
       "1  15540568,15623249,15650814,15659647,15662296,1...           NIC   \n",
       "2  15540568,15623249,15650814,15659647,15662296,1...           FSM   \n",
       "3  15540568,15623249,15650814,15659647,15662296,1...           NIC   \n",
       "4  15540568,15623249,15650814,15659647,15662296,1...           FSM   \n",
       "\n",
       "  novelty_subclasses  GM12878_sum_coverage  K562_sum_coverage  \\\n",
       "0                FSM                     0                226   \n",
       "1  novel combination                     0                 20   \n",
       "2                FSM                     0                 52   \n",
       "3      exon skipping                     0                 20   \n",
       "4                FSM                     0                  6   \n",
       "\n",
       "   GM12878_sum_tpm  K562_sum_tpm  \n",
       "0              0.0   1045.555694  \n",
       "1              0.0     92.527053  \n",
       "2              0.0    240.570337  \n",
       "3              0.0     92.527053  \n",
       "4              0.0     27.758116  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# export transcript table with the same filter criteria:\n",
    "transcript_tab=isoseq.transcript_table( groups=isoseq.groups(),tpm=True,coverage=True,  min_coverage=5, progress_bar=True, query='SUBSTANTIAL and not (NOVEL_TRANSCRIPT and UNSPLICED)')\n",
    "# write to csv file\n",
    "transcript_tab.to_csv(f'{path}/demonstration_dataset_substantial_transcripts.csv.gz', index=False, sep='\\t')\n",
    "#show the first lines\n",
    "transcript_tab.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
