{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ENCODE Long Read Sequel2 dataset\n",
    "\n",
    "This tutorial guides through a typical Long Read Transcriptome Sequencing (LRTS) analysis workflow with isotools, \n",
    "using ENCODE Isoseq SequelII Data. It demonstrates the analysis of alternative splicing events within and between sample groups. \n",
    "\n",
    "## Preparation\n",
    "Select the \"long read RNA-seq\" samples in the ENCODE data portal (https://www.encodeproject.org/) and download aligned .bam files. Here I use all Sequel II samples available at the time of writing, but you choose to process a subset.\n",
    "Download the files with curl and use samtools to index the bam files.\n",
    "``` sh\n",
    "xargs -L 1 curl -O -J -L < ENCODE_files.txt\n",
    "for bam in *bam; \n",
    "    do samtools index $bam\n",
    "done\n",
    "```\n",
    "In addition to the bamfiles, a file metadata.tsv should have been downloaded, which we use to read the sample information. Further you need a reference annotation and a genome fastq file. Please see the Alzheimer tutorial how to get these files. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from  isotools import Transcriptome\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import logging\n",
    "logging.basicConfig(format='%(levelname)s:%(message)s', level=logging.INFO)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>File accession</th>\n",
       "      <th>Output type</th>\n",
       "      <th>Biosample term name</th>\n",
       "      <th>Biosample type</th>\n",
       "      <th>Technical replicate(s)</th>\n",
       "      <th>Platform</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCFF745HHL</td>\n",
       "      <td>alignments</td>\n",
       "      <td>aorta</td>\n",
       "      <td>tissue</td>\n",
       "      <td>1_1</td>\n",
       "      <td>Pacific Biosciences Sequel II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCFF509GHY</td>\n",
       "      <td>alignments</td>\n",
       "      <td>progenitor cell of endocrine pancreas</td>\n",
       "      <td>in vitro differentiated cells</td>\n",
       "      <td>2_1</td>\n",
       "      <td>Pacific Biosciences Sequel II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCFF803KIA</td>\n",
       "      <td>alignments</td>\n",
       "      <td>progenitor cell of endocrine pancreas</td>\n",
       "      <td>in vitro differentiated cells</td>\n",
       "      <td>1_1</td>\n",
       "      <td>Pacific Biosciences Sequel II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCFF971JDY</td>\n",
       "      <td>alignments</td>\n",
       "      <td>mesenteric fat pad</td>\n",
       "      <td>tissue</td>\n",
       "      <td>1_1</td>\n",
       "      <td>Pacific Biosciences Sequel II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCFF292UIE</td>\n",
       "      <td>alignments</td>\n",
       "      <td>H9</td>\n",
       "      <td>cell line</td>\n",
       "      <td>1_1</td>\n",
       "      <td>Pacific Biosciences Sequel II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENCFF738RAA</td>\n",
       "      <td>alignments</td>\n",
       "      <td>H9</td>\n",
       "      <td>cell line</td>\n",
       "      <td>2_1</td>\n",
       "      <td>Pacific Biosciences Sequel II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENCFF911RNV</td>\n",
       "      <td>alignments</td>\n",
       "      <td>right cardiac atrium</td>\n",
       "      <td>tissue</td>\n",
       "      <td>1_1</td>\n",
       "      <td>Pacific Biosciences Sequel II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENCFF472TSL</td>\n",
       "      <td>alignments</td>\n",
       "      <td>posterior vena cava</td>\n",
       "      <td>tissue</td>\n",
       "      <td>1_1</td>\n",
       "      <td>Pacific Biosciences Sequel II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENCFF319JFG</td>\n",
       "      <td>alignments</td>\n",
       "      <td>neural crest cell</td>\n",
       "      <td>in vitro differentiated cells</td>\n",
       "      <td>1_1</td>\n",
       "      <td>Pacific Biosciences Sequel II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENCFF901XCR</td>\n",
       "      <td>alignments</td>\n",
       "      <td>neural crest cell</td>\n",
       "      <td>in vitro differentiated cells</td>\n",
       "      <td>2_1</td>\n",
       "      <td>Pacific Biosciences Sequel II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENCFF810FRP</td>\n",
       "      <td>alignments</td>\n",
       "      <td>HL-60</td>\n",
       "      <td>cell line</td>\n",
       "      <td>1_1</td>\n",
       "      <td>Pacific Biosciences Sequel II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENCFF600MGT</td>\n",
       "      <td>alignments</td>\n",
       "      <td>HL-60</td>\n",
       "      <td>cell line</td>\n",
       "      <td>2_1</td>\n",
       "      <td>Pacific Biosciences Sequel II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENCFF936VUF</td>\n",
       "      <td>alignments</td>\n",
       "      <td>endodermal cell</td>\n",
       "      <td>in vitro differentiated cells</td>\n",
       "      <td>1_1</td>\n",
       "      <td>Pacific Biosciences Sequel II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENCFF939EUU</td>\n",
       "      <td>alignments</td>\n",
       "      <td>endodermal cell</td>\n",
       "      <td>in vitro differentiated cells</td>\n",
       "      <td>2_1</td>\n",
       "      <td>Pacific Biosciences Sequel II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENCFF648NAR</td>\n",
       "      <td>alignments</td>\n",
       "      <td>GM12878</td>\n",
       "      <td>cell line</td>\n",
       "      <td>1_1</td>\n",
       "      <td>Pacific Biosciences Sequel II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENCFF225CCJ</td>\n",
       "      <td>alignments</td>\n",
       "      <td>GM12878</td>\n",
       "      <td>cell line</td>\n",
       "      <td>2_1</td>\n",
       "      <td>Pacific Biosciences Sequel II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENCFF973OML</td>\n",
       "      <td>alignments</td>\n",
       "      <td>type B pancreatic cell</td>\n",
       "      <td>in vitro differentiated cells</td>\n",
       "      <td>2_1</td>\n",
       "      <td>Pacific Biosciences Sequel II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENCFF058HQU</td>\n",
       "      <td>alignments</td>\n",
       "      <td>type B pancreatic cell</td>\n",
       "      <td>in vitro differentiated cells</td>\n",
       "      <td>1_1</td>\n",
       "      <td>Pacific Biosciences Sequel II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENCFF779VVX</td>\n",
       "      <td>alignments</td>\n",
       "      <td>lower lobe of left lung</td>\n",
       "      <td>tissue</td>\n",
       "      <td>1_1</td>\n",
       "      <td>Pacific Biosciences Sequel II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENCFF305AFY</td>\n",
       "      <td>alignments</td>\n",
       "      <td>heart right ventricle</td>\n",
       "      <td>tissue</td>\n",
       "      <td>1_1</td>\n",
       "      <td>Pacific Biosciences Sequel II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENCFF437SYY</td>\n",
       "      <td>alignments</td>\n",
       "      <td>heart left ventricle</td>\n",
       "      <td>tissue</td>\n",
       "      <td>1_1</td>\n",
       "      <td>Pacific Biosciences Sequel II</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   File accession Output type                    Biosample term name  \\\n",
       "0     ENCFF745HHL  alignments                                  aorta   \n",
       "1     ENCFF509GHY  alignments  progenitor cell of endocrine pancreas   \n",
       "2     ENCFF803KIA  alignments  progenitor cell of endocrine pancreas   \n",
       "3     ENCFF971JDY  alignments                     mesenteric fat pad   \n",
       "4     ENCFF292UIE  alignments                                     H9   \n",
       "5     ENCFF738RAA  alignments                                     H9   \n",
       "6     ENCFF911RNV  alignments                   right cardiac atrium   \n",
       "7     ENCFF472TSL  alignments                    posterior vena cava   \n",
       "8     ENCFF319JFG  alignments                      neural crest cell   \n",
       "9     ENCFF901XCR  alignments                      neural crest cell   \n",
       "10    ENCFF810FRP  alignments                                  HL-60   \n",
       "11    ENCFF600MGT  alignments                                  HL-60   \n",
       "12    ENCFF936VUF  alignments                        endodermal cell   \n",
       "13    ENCFF939EUU  alignments                        endodermal cell   \n",
       "14    ENCFF648NAR  alignments                                GM12878   \n",
       "15    ENCFF225CCJ  alignments                                GM12878   \n",
       "16    ENCFF973OML  alignments                 type B pancreatic cell   \n",
       "17    ENCFF058HQU  alignments                 type B pancreatic cell   \n",
       "18    ENCFF779VVX  alignments                lower lobe of left lung   \n",
       "19    ENCFF305AFY  alignments                  heart right ventricle   \n",
       "20    ENCFF437SYY  alignments                   heart left ventricle   \n",
       "\n",
       "                   Biosample type Technical replicate(s)  \\\n",
       "0                          tissue                    1_1   \n",
       "1   in vitro differentiated cells                    2_1   \n",
       "2   in vitro differentiated cells                    1_1   \n",
       "3                          tissue                    1_1   \n",
       "4                       cell line                    1_1   \n",
       "5                       cell line                    2_1   \n",
       "6                          tissue                    1_1   \n",
       "7                          tissue                    1_1   \n",
       "8   in vitro differentiated cells                    1_1   \n",
       "9   in vitro differentiated cells                    2_1   \n",
       "10                      cell line                    1_1   \n",
       "11                      cell line                    2_1   \n",
       "12  in vitro differentiated cells                    1_1   \n",
       "13  in vitro differentiated cells                    2_1   \n",
       "14                      cell line                    1_1   \n",
       "15                      cell line                    2_1   \n",
       "16  in vitro differentiated cells                    2_1   \n",
       "17  in vitro differentiated cells                    1_1   \n",
       "18                         tissue                    1_1   \n",
       "19                         tissue                    1_1   \n",
       "20                         tissue                    1_1   \n",
       "\n",
       "                         Platform  \n",
       "0   Pacific Biosciences Sequel II  \n",
       "1   Pacific Biosciences Sequel II  \n",
       "2   Pacific Biosciences Sequel II  \n",
       "3   Pacific Biosciences Sequel II  \n",
       "4   Pacific Biosciences Sequel II  \n",
       "5   Pacific Biosciences Sequel II  \n",
       "6   Pacific Biosciences Sequel II  \n",
       "7   Pacific Biosciences Sequel II  \n",
       "8   Pacific Biosciences Sequel II  \n",
       "9   Pacific Biosciences Sequel II  \n",
       "10  Pacific Biosciences Sequel II  \n",
       "11  Pacific Biosciences Sequel II  \n",
       "12  Pacific Biosciences Sequel II  \n",
       "13  Pacific Biosciences Sequel II  \n",
       "14  Pacific Biosciences Sequel II  \n",
       "15  Pacific Biosciences Sequel II  \n",
       "16  Pacific Biosciences Sequel II  \n",
       "17  Pacific Biosciences Sequel II  \n",
       "18  Pacific Biosciences Sequel II  \n",
       "19  Pacific Biosciences Sequel II  \n",
       "20  Pacific Biosciences Sequel II  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#I assume the encode data is in an subdirctory \"encode\"\n",
    "\n",
    "#first, process the metadata\n",
    "metadata=pd.read_csv('encode/metadata.tsv', sep='\\t')\n",
    "platform={r['Experiment accession']: r['Platform'] for i,r in metadata[~ metadata.Platform.isnull()].iterrows()}\n",
    "\n",
    "#Select the samples from the metadata (here all SequelII files), make sure there are only bam files are in the table\n",
    "samples=metadata[(metadata['Output type']=='alignments') & (metadata['File Status']=='released')].copy()\n",
    "samples['Platform']=[platform[ea] for ea in samples['Experiment accession']] #This info is missing for some files\n",
    "samples=samples[['File accession','Output type','Biosample term name','Biosample type','Technical replicate(s)','Platform']].reset_index(drop=True) \n",
    "samples=samples[samples.Platform=='Pacific Biosciences Sequel II'].reset_index(drop=True) \n",
    "\n",
    "\n",
    "chrom=[f'chr{i+1}'for i in range(22)]+['chrX','chrY']\n",
    "#ENCODE uses an interesting reference, including sequences from viruses\n",
    "genome='reference/GRCh38_genome_primary_assembly_gencode_v29.fa'\n",
    "anno='reference/gencode.v36.chr_patch_hapl_scaff.annotation_sorted'\n",
    "\n",
    "#print the sample table\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:loading transcriptome from encode/encode_isotools.pkl\n",
      "INFO:importing reference from pkl file reference/gencode.v36.chr_patch_hapl_scaff.annotation_sorted.isotools.pkl\n",
      "INFO:adding sample aorta_1_1\n",
      "100%|██████████| 1960872.0/1960872 [07:11<00:00, 4545.35reads/s, chr=ERCC-00171]             \n",
      "INFO:adding sample progenitor_cell_of_endocrine_pancreas_2_1\n",
      "100%|██████████| 1167968.0/1167968 [04:05<00:00, 4758.71reads/s, chr=ERCC-00171]             \n",
      "INFO:adding sample progenitor_cell_of_endocrine_pancreas_1_1\n",
      "100%|██████████| 1623139.0/1623139 [05:44<00:00, 4708.39reads/s, chr=ERCC-00171]             \n",
      "INFO:adding sample mesenteric_fat_pad_1_1\n",
      "100%|██████████| 1816266.0/1816266 [04:50<00:00, 6253.37reads/s, chr=ERCC-00171]             \n",
      "INFO:adding sample H9_1_1\n",
      "100%|██████████| 3201257.0/3201257 [10:35<00:00, 5036.39reads/s, chr=ERCC-00171]             \n",
      "INFO:adding sample H9_2_1\n",
      "100%|██████████| 2753910.0/2753910 [08:37<00:00, 5325.41reads/s, chr=ERCC-00171]             \n",
      "INFO:adding sample right_cardiac_atrium_1_1\n",
      "100%|██████████| 2392477.0/2392477 [08:59<00:00, 4435.22reads/s, chr=ERCC-00171]             \n"
     ]
    }
   ],
   "source": [
    "try:    \n",
    "    isoseq=Transcriptome(f'encode/encode_isotools.pkl')\n",
    "except FileNotFoundError:\n",
    "    try:\n",
    "        isoseq=Transcriptome.from_reference(anno+'.isotools.pkl')\n",
    "    except FileNotFoundError:\n",
    "        isoseq=Transcriptome.from_reference(anno+'.gff3.gz')\n",
    "        isoseq.save_reference(anno+'.isotools.pkl')\n",
    "    for i,row in samples.iterrows():\n",
    "        sname=f\"{row['Biosample term name']}_{row['Technical replicate(s)']}\".replace(' ','_')\n",
    "        if sname in isoseq.samples:\n",
    "            sname+='_b'\n",
    "        isoseq.add_sample_from_bam(f\"encode/{row['File accession']}.bam\", \n",
    "                                   sample_name=sname , \n",
    "                                   group=row['Biosample term name'] ,\n",
    "                                   sample_type=row['Biosample type'],\n",
    "                                   platform=row['Platform']) \n",
    "    isoseq.add_biases(genome_fn) #3:41 min\n",
    "    isoseq.make_index()\n",
    "    isoseq.save('encode/encode_isotools.pkl')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Quality control and filtering\n",
    "First, we will depict some quality metrics and define filtering criteria."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#add gencode specific filters\n",
    "ref_filter={'UNSPLICED':'len(exons)==1','HIGH_SUPPORT':'transcript_support_level==\"1\"', 'PROTEIN_CODING':'transcript_type==\"protein_coding\"'}\n",
    "#encode bams do not have chimeric redas \n",
    "isoseq.add_filter( gene_filter=gene_filter, ref_transcript_filter=ref_filter)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute some summary statistics on technical artifacts. \n",
    "# For this analysis samples are grouped by 'Biosample term name'\n",
    "tr_stats=[\n",
    "    isoseq.transcript_length_hist(groups=isoseq.groups(), add_reference=True, min_coverage=2,tr_filter=dict( remove=['NOVEL_GENE']), ref_filter=dict(include=['HIGH_SUPPORT'])),\n",
    "    isoseq.downstream_a_hist(groups=isoseq.groups(), tr_filter=dict( remove=['NOVEL_GENE', 'UNSPLICED']), ref_filter=dict(remove=['UNSPLICED'])),\n",
    "    isoseq.downstream_a_hist(groups=isoseq.groups(), tr_filter=dict(include=['NOVEL_GENE', 'UNSPLICED'])),\n",
    "    isoseq.direct_repeat_hist(groups=isoseq.groups(),bins=np.linspace(-.5,10.5,12))]\n",
    "\n",
    "tr_stats.append((pd.concat([tr_stats[2][0].add_suffix(' novel unspliced'),tr_stats[1][0].add_suffix(' known multiexon')],axis=1),tr_stats[2][1]))\n",
    "\n",
    "#statistic on the filter flags\n",
    "f_stats=isoseq.filter_stats( groups=isoseq.groups(), weight_by_coverage=True,min_coverage=1)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Exploration\n",
    "To explore the relation of the samples with respect to splicing we look at PCA and UMAP embeddings based on alternative splicing events."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Compute alternative splicing events by finding \"bubbles\" in the segment graph\n",
    "splice_events=isoseq.find_splice_bubbles()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from isotools.plots import plot_embedding\n",
    "plt.rcParams[\"figure.figsize\"] = (20,10)\n",
    "components={}\n",
    "pca={}\n",
    "p={}\n",
    "f,axs=plt.subplots(2,4)\n",
    "for ax,t in zip(axs.flatten(),['all','3AS','5AS','ES','IR','ME', 'TSS', 'PAS']):\n",
    "    p_i,comp_i, pca_i=plot_embedding(splice_events, ax=ax, labels=True, groups=isoseq.groups(), splice_types=t)#, colors={'mut':'red', 'wt':'green'})\n",
    "    components[t]=comp_i\n",
    "    pca[t]=pca_i\n",
    "    p[t]=p_i\n",
    "\n",
    "axs[0,2].legend(fontsize='medium', ncol=1,handleheight=2.4, labelspacing=0.05, bbox_to_anchor=(1.02, 1), loc='upper left')\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(out_fn+'_PCA.png') \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
